packages/loader/container-loader/src/container.ts

```diff
@@ -379,81 +379,104 @@ export class Container
-
-					container
-						.load(version, mode, resolvedUrl, pendingLocalState)
-						.finally(() => {
-							container.removeListener("closed", onClosed);
-						})
-						.then(
-							(props) => {
-								event.end({ ...props });
-								resolve(container);
-							},
-							(error) => {
-								const err = normalizeError(error);
-								// Depending where error happens, we can be attempting to connect to web socket
-								// and continuously retrying (consider offline mode)
-								// Host has no container to close, so it's prudent to do it here
-								// Note: We could only dispose the container instead of just close but that would
-								// the telemetry where users sometimes search for ContainerClose event to look
-								// for load failures. So not removing this at this time.
-								container.close(err);
-								container.dispose(err);
-								onClosed(err);
-							},
-						);
-				}),
-			{ start: true, end: true, cancel: "generic" },
-		);
+		const initialize = async (): Promise<void> =>
+			PerformanceEvent.timedExecAsync(
+				container.mc.logger,
+				{ eventName: "Load", ...loadMode },
+				async (event) =>
+					new Promise((resolve, reject) => {
+						const defaultMode: IContainerLoadMode = { opsBeforeReturn: "cached" };
+						// if we have pendingLocalState, anything we cached is not useful and we shouldn't wait for connection
+						// to return container, so ignore this value and use undefined for opsBeforeReturn
+						const mode: IContainerLoadMode = pendingLocalState
+							? { ...(loadMode ?? defaultMode), opsBeforeReturn: undefined }
+							: (loadMode ?? defaultMode);
+
+						const onClosed = (err?: ICriticalContainerError): void => {
+							// pre-0.58 error message: containerClosedWithoutErrorDuringLoad
+							reject(err ?? new GenericError("Container closed without error during load"));
+						};
+						container.on("closed", onClosed);
+
+						container
+							.load(version, mode, resolvedUrl, pendingLocalState)
+							.finally(() => {
+								container.removeListener("closed", onClosed);
+							})
+							.then(
+								(props) => {
+									event.end({ ...props });
+									resolve();
+								},
+								(error) => {
+									const err = normalizeError(error);
+									// Depending where error happens, we can be attempting to connect to web socket
+									// and continuously retrying (consider offline mode)
+									// Host has no container to close, so it's prudent to do it here
+									// Note: We could only dispose the container instead of just close but that would
+									// the telemetry where users sometimes search for ContainerClose event to look
+									// for load failures. So not removing this at this time.
+									container.close(err);
+									container.dispose(err);
+									onClosed(err);
+								},
+							);
+					}),
+				{ start: true, end: true, cancel: "generic" },
+			);
+
+		return Object.assign(container, { initialize });
```

Comment:
i really don't like the use of object.assign in this pr. ideally we can decompose whatever is happeing into free functions and small objects to build up the container, rather than tacking new functions on top this Container monolith

Reply:
Yeah I don't either, just playing around.  I struggle to communicate what I'm looking for feedback on sometimes with these exploratory PRs.